### ISSUE TITLE
Use IPX as high performance, secure and easy-to-use image optimizer.

### ISSUE THREAD

**Issue by Murderlon** [rehypejs/camomile#2]
### Initial checklist

- [X] I read the [support docs](https://github.com/rehypejs/.github/blob/main/support.md)
- [X] I read the [contributing guide](https://github.com/rehypejs/.github/blob/main/contributing.md)
- [X] I agree to follow the [code of conduct](https://github.com/rehypejs/.github/blob/main/code-of-conduct.md)
- [X] I searched [issues](https://github.com/search?q=user%3Arehypejs&type=issues) and [discussions](https://github.com/orgs/rehypejs/discussions) and couldn’t find anything (or linked relevant results below)

### Problem

One thing that could be expected/wanted from an image proxy is also optimizing images and/or changing the image format.

### Current solutions

This was already on our minds but building it yourself from scratch is a lot of work.

### Proposed solutions

unjs is a great ecosystem focussed on universal tooling, much like unified's philosophy. They have an image optimizer which can be easily integrated into Node.js servers.

https://github.com/unjs/ipx

**Comment by ChristianMurphy** [comment]
Sounds interesting!
Worth looking into.

One consideration unpacking images opens up a resource exhaustion risk.
For example https://www.bamsoftware.com/hacks/deflate.html

It could be good to have defaults that mitigate these risks.

**Comment by wooorm** [comment]
I worry a bit about using a tool that wraps other tools.
Especially but not solely for security stuff like what Christian mentions.

If the proposed solution is already very simple to integrate into a server, then what is missing?
Would an example showing how to use 2 tools together solve this issue?

**Comment by Murderlon** [comment]
> If the proposed solution is already very simple to integrate into a server, then what is missing?

Briefly looking at it again, ipx seems to be mostly designed to be used as req/res middleware. However that doesn't work in our case, even if we introduce `camomile.use()`, as we need it to operate on the result of the fetch, not the original request.

So we'd have to add something like `options.onBufferLoaded(buffer: Buffer): Buffer` and add an example with a lower level thing from ipx, not the request handler, to transform the image. This works because we actually buffer the whole image into memory before moving on.

**Comment by wooorm** [comment]
sharp allows many manipulations, and ipx seems to allow that too. So that’s a bit more complex.
But svgo is more simple (although: svg is also unsafe, not sure we should support it? I don’t think GH does normally either).
But, we could have examples such as this in the readme here?https://github.com/unjs/ipx/blob/f8455cd466aef067f8645ebfa592f57449811532/src/ipx.ts#L170-L191

**Comment by Murderlon** [comment]
> So that’s a bit more complex

Yeah in which case I would say building it inside camomile with an opt-in option still isn't a bad idea, especially since we're tightly coupled to formats we support. If we decide we want it, and we determine the code is non-trivial (too big for an example), then it's better to figure it out with security in mind once instead of pushing the risk to implementers.

**Comment by wooorm** [comment]
What is the `it`? sharp? Or ipx, or a hook?

The `it` I’d lean to would be an example for sharp, plus hooks in the API where needed to make the example work. From what I understand sharp + 20 lines of example code + 1 hook could solve this issue?

Indeed, if it is non-trivial, and we can make it secure, optionally with ipx, then we could expose an implementation of that hook in the future?

The hook would be `options.onBufferLoaded` you talk about. From what I see in the code, and given the use case, I’m thinking it would be [called here](https://github.com/rehypejs/camomile/blob/e7efb1422ca8f208d9a387bc36e1b289df28f2de/lib/server.js#L211). For a name, perhaps `options.transform` or `transformImage` or `transformResponse` might be better? That hook might also want to get the mime type we already inferred and checked? Perhaps other headers might be useful for sharp/svgo?

**Comment by Murderlon** [comment]
> The `it` I’d lean to would be an example for sharp, plus hooks in the API where needed to make the example work. From what I understand sharp + 20 lines of example code + 1 hook could solve this issue?

Yes that sounds good.
